internal/assertions/apiversions_response_assertion.go

```diff
@@ -29,51 +20,99 @@ var errorCodes = map[int]string{
-					}
-					logger.Successf("✓ MaxVersion for %s is >= %v", apiKeyNames[expectedApiVersionKey.ApiKey], expectedApiVersionKey.MaxVersion)
+	return a
+}
+
+// assertAPIKeysArray asserts the API keys array in the response
+// Fields asserted by default: ApiKeysArray.length, ApiKeyArrayElement.MinVersion, ApiKeyArrayElement.MaxVersion
+func (a *ApiVersionsResponseAssertion) assertAPIKeysArray() *ApiVersionsResponseAssertion {
+	if a.err != nil {
+		return a
+	}
+
+	if len(a.ActualValue.ApiKeys) < len(a.ExpectedValue.ApiKeys) {
+		a.err = fmt.Errorf("Expected API keys array to include atleast %d keys, got %d", len(a.ExpectedValue.ApiKeys), len(a.ActualValue.ApiKeys))
+		return a
+	}
+	a.logger.Successf("✓ API keys array length: %d", len(a.ActualValue.ApiKeys))
+
+	for _, expectedApiVersionKey := range a.ExpectedValue.ApiKeys {
+		found := false
+		for _, actualApiVersionKey := range a.ActualValue.ApiKeys {
+			if actualApiVersionKey.ApiKey == expectedApiVersionKey.ApiKey {
+				found = true
+				if actualApiVersionKey.MinVersion > expectedApiVersionKey.MaxVersion {
+					a.err = fmt.Errorf("Expected min version %v to be < max version %v for %s", actualApiVersionKey.MinVersion, expectedApiVersionKey.MaxVersion, apiKeyNames[expectedApiVersionKey.ApiKey])
+					return a
 				}
-			}
-			if !found {
-				return fmt.Errorf("Expected APIVersionsResponseKey array to include API key %d (%s)", expectedApiVersionKey.ApiKey, apiKeyNames[expectedApiVersionKey.ApiKey])
+
+				// anything above or equal to expected minVersion is fine
+				if actualApiVersionKey.MinVersion < expectedApiVersionKey.MinVersion {
+					a.err = fmt.Errorf("Expected API version %v to be supported for %s, got %v", expectedApiVersionKey.MaxVersion, apiKeyNames[expectedApiVersionKey.ApiKey], actualApiVersionKey.MaxVersion)
+					return a
+				}
+				a.logger.Successf("✓ MinVersion for %s is <= %v & >= %v", apiKeyNames[expectedApiVersionKey.ApiKey], expectedApiVersionKey.MaxVersion, expectedApiVersionKey.MinVersion)
+
+				if actualApiVersionKey.MaxVersion < expectedApiVersionKey.MaxVersion {
+					a.err = fmt.Errorf("Expected API version %v to be supported for %s, got %v", expectedApiVersionKey.MaxVersion, apiKeyNames[expectedApiVersionKey.ApiKey], actualApiVersionKey.MaxVersion)
+					return a
+				}
+				a.logger.Successf("✓ MaxVersion for %s is >= %v", apiKeyNames[expectedApiVersionKey.ApiKey], expectedApiVersionKey.MaxVersion)
 			}
 		}
+		if !found {
+			a.err = fmt.Errorf("Expected APIVersionsResponseKey array to include API key %d (%s)", expectedApiVersionKey.ApiKey, apiKeyNames[expectedApiVersionKey.ApiKey])
+			return a
+		}
 	}
 
-	return nil
+	return a
+}
+
+// Run runs assertBody and assertAPIKeysArray in sequence
+func (a *ApiVersionsResponseAssertion) Run() error {
+	a.assertBody()
+	a.assertAPIKeysArray()
+
+	return a.err
```

Comment:
We aren't chaining anymore, so there's no reason to store `err` as state on the struct. We should just do what we'd usually do when running multiple methods in serial and returning errors: 

```
if err := a.assertBody(); err != nil { 
  return err 
}

if err := a.assertAPIKeysArray... {
  return err 
}
```